src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/DiagnosticOr.cs

```diff
@@ -0,0 +1,146 @@
+            return (values, diagnostics);
+        }
+
+        /// <summary>
+        /// Splits the elements of <paramref name="provider"/> into groups depending on whether Item1 is a value or a Diagnostic.
+        /// </summary>
+        public static (IncrementalValuesProvider<(T, T2)>, IncrementalValuesProvider<Diagnostic>) Split<T, T2>(this IncrementalValuesProvider<(DiagnosticOr<T>, T2)> provider)
+        {
+            var values = provider.Where(x => x.Item1.IsValue).Select(static (x, ct) => (x.Item1.Value, x.Item2));
+            var diagnostics = provider.Where(x => !x.Item1.IsValue).Select(static (x, ct) => x.Item1.Diagnostic);
+            return (values, diagnostics);
+        }
+    }
+
+    public static partial class IncrementalGeneratorInitializationContextExtensions
+    {
+        /// <summary>
+        /// Filters the <see cref="IncrementalValuesProvider{TValue}"/> by whether or not the is a <see cref="Diagnostic"/>, reports the diagnostics, and returns the values.
+        /// </summary>
+        public static IncrementalValuesProvider<T> FilterAndReportDiagnostics<T>(this IncrementalGeneratorInitializationContext ctx, IncrementalValuesProvider<DiagnosticOr<T>> diagnosticOrValues)
+        {
+            var (values, diagnostics) = diagnosticOrValues.Split();
+            ctx.RegisterDiagnostics(diagnostics);
+            return values;
+        }
+
+        /// <summary>
+        /// Filters both <see cref="IncrementalValuesProvider{TValue}"/> by whether or not the value in <paramref name="diagnosticOrValues"/> is a <see cref="Diagnostic"/>, reports the diagnostics, and returns the values.
+        /// </summary>
+        public static (IncrementalValuesProvider<T>, IncrementalValuesProvider<T2>) FilterAndReportDiagnostics<T, T2>(
+            this IncrementalGeneratorInitializationContext ctx,
+            IncrementalValuesProvider<DiagnosticOr<T>> diagnosticOrValues,
+            IncrementalValuesProvider<T2> associatedValues)
+        {
+            var (values, diagnostics) = diagnosticOrValues.Zip(associatedValues).Split();
+            ctx.RegisterDiagnostics(diagnostics);
+            return values.Split();
+        }
+
+        /// <summary>
+        /// Filters each inner <see cref="IEnumerable{T}"/> of <see cref="DiagnosticOr{T}"/> by whether the elements are <see cref="Diagnostic"/>s, reports the diagnostics, and returns the values.
+        /// </summary>
+        public static IncrementalValuesProvider<SequenceEqualImmutableArray<T>> FilterAndReportDiagnostics<T, TEnumerable>(
+            this IncrementalGeneratorInitializationContext ctx,
+            IncrementalValuesProvider<TEnumerable> diagnosticOrValues)
+            where TEnumerable : IEnumerable<DiagnosticOr<T>>
+        {
+            var (values, diagnostics) = diagnosticOrValues.SplitArrays<T, TEnumerable>();
+            ctx.RegisterDiagnostics(diagnostics);
+            return values;
+        }
```

Comment:
Can we come up with an API shape for `FilterAndReportDiagnostics` that doesn't require us to specify the generic type parameters at the use site?

Reply:
Closed TEnumerable to SequenceEqualImmutableArray<T>